src/systems/camera-system.js

```diff
@@ -0,0 +1,182 @@
+const m3 = new THREE.Matrix4();
+const offset = new THREE.Vector3();
+export class CameraSystem {
+  constructor() {
+    this.mode = CAMERA_MODE_FIRST_PERSON;
+    waitForDOMContentLoaded().then(() => {
+      this.playerCamera = document.getElementById("player-camera");
+      this.playerRig = document.getElementById("player-rig");
+      this.cameraEl = document.getElementById("experimental-camera");
+      this.rigEl = document.getElementById("experimental-rig");
+      this.rigEl.object3D.matrixWorldNeedsUpdate = true;
+      this.rigEl.object3D.matrixIsModified = true;
+    });
+  }
+  nextMode() {
+    if (!qsTruthy("thirdPerson")) return;
+
+    if (this.mode === CAMERA_MODE_INSPECT) {
+      this.uninspect();
+      return;
+    }
+
+    this.mode = (this.mode + 1) % numCameraModes;
+    if (this.mode === CAMERA_MODE_FIRST_PERSON) {
+      AFRAME.scenes[0].renderer.vr.setPoseTarget(this.playerCamera.object3D);
+    } else if (this.mode === CAMERA_MODE_THIRD_PERSON_NEAR || this.mode === CAMERA_MODE_THIRD_PERSON_FAR) {
+      AFRAME.scenes[0].renderer.vr.setPoseTarget(this.cameraEl.object3D);
+    } else if (this.mode === CAMERA_MODE_INSPECT) {
+      this.mode = CAMERA_MODE_FIRST_PERSON; // skip inspect for now
+    }
+  }
+  inspect(o) {
+    if (this.mode !== CAMERA_MODE_INSPECT) {
+      this.preInspectMode = this.mode;
+    }
+    this.mode = CAMERA_MODE_INSPECT;
+    this.inspected = o;
+    moveRigSoCameraLooksAtObject(this.rigEl.object3D, this.cameraEl.object3D, this.inspected);
+  }
+  uninspect() {
+    this.inspected = null;
+    if (this.mode !== CAMERA_MODE_INSPECT) {
+      return;
+    }
+    this.mode = this.preInspectMode || CAMERA_MODE_FIRST_PERSON;
+  }
+  tick() {
+    this.playerHead = this.playerHead || document.getElementById("player-head");
+    if (!this.playerHead) return;
+    if (!this.playerCamera) return;
+    this.playerCamera.components["pitch-yaw-rotator"].on = true;
+    this.cameraEl.components["pitch-yaw-rotator"].on = true;
+
+    const userinput = AFRAME.scenes[0].systems.userinput;
+    if (userinput.get(paths.actions.nextCameraMode)) {
+      this.nextMode();
+    }
+
+    const visible = this.mode !== CAMERA_MODE_FIRST_PERSON;
+    if (visible !== this.playerHead.object3D.visible) {
+      this.playerHead.object3D.visible = visible;
+    }
+
+    this.playerRig.object3D.updateMatrices();
+    this.rigEl.object3D.updateMatrices();
+    if (this.mode === CAMERA_MODE_FIRST_PERSON) {
+      this.cameraEl.components["pitch-yaw-rotator"].on = false;
+      m2.copy(this.playerRig.object3D.matrixWorld);
+      setMatrixWorld(this.rigEl.object3D, m2);
+    }
+
+    this.playerCamera.object3D.matrixWorldNeedsUpdate = true;
+    if (this.mode === CAMERA_MODE_FIRST_PERSON) {
+      if (!AFRAME.scenes[0].is("vr-mode")) {
+        setMatrixWorld(this.cameraEl.object3D, this.playerCamera.object3D.matrixWorld);
+      } else {
+        setMatrixWorld(this.playerCamera.object3D, this.cameraEl.object3D.matrixWorld);
```

Comment:
`this.cameraEl.object3D.updateMatrices()` should be called before reading `matrixWorld` here